src/visual.ts

```diff
@@ -330,64 +343,80 @@ module powerbi.extensibility.visual {
         }
 
         private createNodes(
-            categories: any[],
+            sourceCategories: any[],
+            destinationCategories: any[],
             settings: SankeyDiagramSettings,
             selectionIdBuilder: SankeyDiagramSelectionIdBuilder,
-            source: DataViewMetadataColumn): SankeyDiagramNode[] {
+            source: DataViewMetadataColumn,
+            linksObjects: DataViewObjects[],
+            sourceCategoriesLabels?: any[],
+            destinationCategoriesLabels?: any[]): SankeyDiagramNode[] {
 
             let nodes: SankeyDiagramNode[] = [],
                 valueFormatterForCategories: IValueFormatter;
 
             valueFormatterForCategories = ValueFormatter.create({
                 format: ValueFormatter.getFormatStringByColumn(source),
-                value: categories[0],
-                value2: categories[categories.length - 1]
+                value: sourceCategories[0],
+                value2: destinationCategories[destinationCategories.length - 1]
             });
 
-            categories.forEach((item: any, index: number) => {
-                if (!nodes.some((node: SankeyDiagramNode) => {
-                    if (item === node.label.name) {
-                        const selectionId: ISelectionId = selectionIdBuilder.createSelectionId(index);
+            // check self connected links
+            for (let index: number = 0; index < destinationCategories.length; index++) {
+                if (sourceCategoriesLabels[index] === undefined) {
+                    sourceCategoriesLabels[index] = sourceCategories[index];
+                }
+                if (destinationCategoriesLabels[index] === undefined) {
+                    destinationCategoriesLabels[index] = destinationCategories[index];
+                }
+                if (sourceCategories[index] === destinationCategories[index]) {
+                    destinationCategories[index] += "_SK_SELFLINK";
```

Comment:
Why can't we rewrite design of the visual?
If it isn't possible please move the string to a variable. It would allow us to modify this value at one place.